### ISSUE TITLE
Code Review Session 1

### ISSUE THREAD

**Issue by LKKTWH** [michellekappl/Preprocessing_Pipeline_JuML#14]
**1.	Write down your “user experience”:**

**- Which functionality was usable?**
The package provides important decision tree related operations. ①construction of classification trees (_DecisionTreeClassifier_) ②Construction of regression trees(_DecisionTreeRegressor_) ③Prediction function(_predict_), and so on. Also functions related with visualization and debugging are provided: ①Visualization of tree structure (_print_tree_) ②Depth calculation (_calc_depth_)

**- Was the package intuitive to use?** 
Yes. ①The API design is clear. The ‘fit!’ and ‘_predict_’ paradigms are implemented and also often seen in machine learning libraries, ②Error Handling is strong. The input is validated, and clear error messages with enough information can help users fast identify the problem.

**- What was documented?** 
①	Parameter Information: input params for every function are clearly listed. Return values with their types, and expected data types and formats are also specified.
②	Instructions on how to use the package: there are examples showing how to utilize the functions. Guidelines for many different features also explicitly mentioned. 

**- How well was it documented?** 
①Comprehensive： Each function has detailed explanations of its purpose and behavior. Documentation includes both basic usage and edge cases
②Clear： Documentation follows a consistent format across the implementation. Examples are provided to demonstrate the correct usage
③Usable practically: Users are guided how to use the package correctly by the given examples and documentations. Error situations are also specified.

**- Did you encounter any errors?**
For the ‘OneTwoTree’ the package itself, No. Only some Julia environment setting problem when installing the package, but Tistou and me fixed it.

**2. Are there any issues with the code? / Do you see ways to improve the performance?**

a)	The hyperparameter tuning is limited for the decision tree, e.g. only considered ‘max_depth’ to decide whether the tree to split or not. Maybe consider adding other parameters (reference: https://scikit-learn.org/1.5/modules/generated/sklearn.ensemble.RandomForestClassifier.html)

b)	Lack of data pre-processing. The datasets may contain null values of NA. Users may have to make a decision about whether to remove these entries or fill them with some values. The lack of functions like fillna(), handle_missing() forces users to write functions to preprocess their datasets before actually using your decision tree/random forests functions. Also data standardization and normalization are crucial for pre-processing data, since different features within a dataset might have different scales.

c)     Also there should be more different types of test dataset coverage. 

**3. Is the source code well structured?**
Yes, the source code is well structured. Modular Organized, Clear Abstraction Hierarchy, Consistent coding style through the whole implementation. 

**4. Are there features you would like to see?**
①	Pruning to prevent overfitting ② Parallel processing for large datasets ③ Outlier detection and handling ④ Support for weighted samples

**5. Is the well tested? Do the tests look trustworthy?**
For the basic function implemented until now, the given test well tested the functions. Maybe the tests could cover more outlier siturations.
                                                                                                                                   (Ran Zhang)

**Comment by leonschreiber96** [comment]
This has nothing to do at all with the code or the purpose of this package. The pipeline contains neither any decision trees, nor actual ML functionality like `fit()`or `predict()` methods (and never will). It is a preprocessing pipeline for NLP datasets.

Also, GitHub issues are for pinpointing certain concrete flaws, bugs or feature requests, not general code reviews. In this text I don't see any of that, except for vague feature idas that have nothing to do with the package purpose. Therefore, I will close it without further action.
